common/flatpak-dir.c

```diff
@@ -7652,31 +7694,87 @@ flatpak_dir_install (FlatpakDir          *self,
+              g_autoptr(OstreeRepo) repo = ostree_repo_new (repo_file);
+              if (!ostree_repo_open (repo, NULL, &local_error))
+                {
+                  g_warning ("Failed to open child repo at %s: %s", helper_child_repo_path, local_error->message);
+
+                  /* Explicitly clearing the child_repo pointer closes any open FDs on the child_repo.
+                     This prevents EBUSY error failure during unmounting. */
+                  g_clear_object (&repo);
+                  flatpak_dir_cancel_pull (self, cancellable, helper_child_repo_path);
+                }
+              else
+                child_repo = g_steal_pointer (&repo);
+            }
+
           /* We're pulling from a remote source, we do the network mirroring pull as a
              user and hand back the resulting data to the system-helper, that trusts us
              due to the GPG signatures in the repo */
 
-          child_repo = flatpak_dir_create_system_child_repo (self, &child_repo_lock, NULL, error);
+          /* Fallback */
           if (child_repo == NULL)
-            return FALSE;
+            {
+              /* We're pulling from a remote source, we do the network mirroring pull as a
+                 user and hand back the resulting data to the system-helper, that trusts us
+                 due to the GPG signatures in the repo */
+              child_repo = flatpak_dir_create_system_child_repo (self, &child_repo_lock, NULL, error);
+              if (child_repo == NULL)
+                return FALSE;
+
+              fallback = TRUE;
+            }
+           child_repo_path = g_file_get_path (ostree_repo_get_path (child_repo));
 
           flatpak_flags |= FLATPAK_PULL_FLAGS_SIDELOAD_EXTRA_DATA;
 
           /* Don’t resolve a rev or OstreeRepoFinderResult set early; the pull
            * code will do this. */
-          if (!flatpak_dir_pull (self, state, ref, opt_commit, NULL, subpaths,
-                                 child_repo,
-                                 flatpak_flags,
-                                 OSTREE_REPO_PULL_FLAGS_MIRROR,
-                                 progress, cancellable, error))
-            return FALSE;
-
-          if (!child_repo_ensure_summary (child_repo, state, cancellable, error))
-            return FALSE;
-
-          child_repo_path = g_file_get_path (ostree_repo_get_path (child_repo));
+          if (!flatpak_dir_pull (self, state, ref, opt_commit, NULL, subpaths, child_repo,
+                                 flatpak_flags, OSTREE_REPO_PULL_FLAGS_MIRROR,
+                                 progress, cancellable, error)
+              || !child_repo_ensure_summary (child_repo, state, cancellable, error))
+            {
+              if (!fallback)
+                {
+                  /* Explicitly clearing the child_repo pointer closes any open FDs on the child_repo.
+                     This prevents EBUSY error failure during unmounting. */
```

Comment:
Is there a chance that the client can dos the system-helper here by keeping an fd open, preventing the bindfs from being undone ?

Reply:
Yes, it can. This is what we are trying to avoid here.

If we can't umount cleanly(or DOS ?), we then try to `force_exit` the bindfs subprocess.